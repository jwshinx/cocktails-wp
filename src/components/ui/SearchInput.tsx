// used by cocktails
import React, { useState, useEffect } from 'react'

import useDebounce from '../../hooks/useDebounce'
import { SegmentKeyType } from '../../models/cocktail'

interface SearchInputProps {
  segmentType?: string
  segmentKey?: SegmentKeyType
  setSearchQuery: (searchQuery: string) => void
  placeholderText: string
}

export function SearchInput(props: SearchInputProps) {
  const { setSearchQuery, placeholderText } = props
  const [query, setQuery] = useState<string>('')

  let isOnlyOneLetterError = false
  if (
    props.segmentType === 'search' &&
    props.segmentKey?.name === 'firstLetter' &&
    query.length > 1
  ) {
    isOnlyOneLetterError = true
  }

  const debouncedQuery = useDebounce(query, 500)

  useEffect(() => {
    setSearchQuery(debouncedQuery)
  }, [debouncedQuery, setSearchQuery])

  return (
    <div className="form-group">
      <input
        id="search"
        type="search"
        className="form-control"
        placeholder={placeholderText}
        onChange={(event) => setQuery(event.target.value)}
      />
      <small
        id="error"
        className={`form-text ${
          isOnlyOneLetterError ? 'text-danger' : 'd-none'
        }`}
      >
        Please enter single letter.
      </small>
    </div>
  )
}
