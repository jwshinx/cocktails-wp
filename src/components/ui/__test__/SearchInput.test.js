/**
 * @jest-environment jsdom
 */

import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'

import { SearchInput } from '../SearchInput'

const mockFn = jest.fn(() => true)

it('should render searchbox with "Enter a value" placeholder', () => {
  render(
    <SearchInput placeholderText="Enter a value" setSearchQuery={mockFn} />
  )
  const elemByPlaceholder = screen.getByPlaceholderText(/Enter a value/)
  expect(elemByPlaceholder).toBeInTheDocument()

  const elemByRole = screen.getByRole('searchbox')
  expect(elemByRole).toBeInTheDocument()
})

it('should snapshot SearchInput', () => {
  const tree = renderer
    .create(
      <SearchInput placeholderText="Enter a value" setSearchQuery={mockFn} />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
