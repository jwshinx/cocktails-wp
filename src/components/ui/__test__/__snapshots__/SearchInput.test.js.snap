// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should snapshot SearchInput 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control"
    id="search"
    onChange={[Function]}
    placeholder="Enter a value"
    type="search"
  />
  <small
    className="form-text d-none"
    id="error"
  >
    Please enter single letter.
  </small>
</div>
`;
